akka-contrib/src/test/scala/akka/contrib/circuitbreaker/CircuitBreakerActorSpec.scala

```diff
@@ -0,0 +1,302 @@
+      receiver.expectNoMsg()
+
+      sender.expectMsg(CircuitOpenFailure("Second message in half-open state, should be ignored"))
+
+    }
+
+    "return to CLOSED state from HALF-OPEN if a successful message response notification is received" in new CircuitBreakerScenario {
+      Given("A circuit breaker actor pointing to a test probe")
+      val circuitBreaker = defaultCircuitBreaker
+
+      When("Entering HALF OPEN state")
+      receiverRespondsWithFailureToRequest("request1")
+      receiverRespondsWithFailureToRequest("request2")
+
+      waitForResetTimeoutToExpire
+
+      And("Receiving a successful response")
+      receiverRespondsToRequestWith("First message in half-open state, should be forwarded", "This should close the circuit")
+
+      waitForCircuitBreakerToReceiveSelfNotificationMessage
+
+      Then("circuit is re-closed")
+      sender.send(circuitBreaker, "request1")
+      receiver.expectMsg("request1")
+
+      sender.send(circuitBreaker, "request2")
+      receiver.expectMsg("request2")
+
+    }
+
+    "return to OPEN state from HALF-OPEN if a FAILURE message response is received" in new CircuitBreakerScenario {
+      Given("A circuit breaker actor pointing to a test probe")
+      val circuitBreaker = defaultCircuitBreaker
+
+      When("Entering HALF OPEN state")
+      receiverRespondsWithFailureToRequest("request1")
+      receiverRespondsWithFailureToRequest("request2")
+
+      waitForResetTimeoutToExpire
+
+      And("Receiving a failure response")
+      receiverRespondsWithFailureToRequest("First message in half-open state, should be forwarded")
+
+      waitForCircuitBreakerToReceiveSelfNotificationMessage
+
+      Then("circuit is opened again")
+      sender.send(circuitBreaker, "this should be ignored")
+      receiver.expectNoMsg()
+      sender.expectMsg(CircuitOpenFailure("this should be ignored"))
+
+    }
+
+    "Notify an event status change listener when changing state" in new CircuitBreakerScenario {
+      Given("A circuit breaker actor pointing to a test probe")
+      override val circuitBreaker = system.actorOf(
+        baseCircuitBreakerBuilder
+          .copy(circuitEventListener = Some(eventListener.ref))
+          .propsForTarget(receiver.ref))
+
+      When("Entering OPEN state")
+      receiverRespondsWithFailureToRequest("request1")
+      receiverRespondsWithFailureToRequest("request2")
+
+      waitForCircuitBreakerToReceiveSelfNotificationMessage
+
+      Then("An event is sent")
+      eventListener.expectMsg(CircuitOpen(circuitBreaker))
+
+      When("Entering HALF OPEN state")
+      waitForResetTimeoutToExpire
+
+      Then("An event is sent")
+      eventListener.expectMsg(CircuitHalfOpen(circuitBreaker))
+
+      When("Entering CLOSED state")
+      receiverRespondsToRequestWith("First message in half-open state, should be forwarded", "This should close the circuit")
+      Then("An event is sent")
+      eventListener.expectMsg(CircuitClosed(circuitBreaker))
+
+    }
```

Comment:
needs a test (and impl, since it does not yet do that) for this scenario:
- circuit breaker forwards a message which causes the actor to stop
- circuit breaker should stop as well, since it'll now just keep sending into deadLetters otherwise.

This can be solved by doing a `context.watch(target)` in the `preStart` of the CircuitBreakerProxy actor.